README
NAME
    Image::Scale - Fast, high-quality fixed-point image resizing

SYNOPSIS
        use Image::Scale;
    
        # Resize to 150 width and save to a file
        my $img = Image::Scale->new('image.jpg') || die "Invalid JPEG file";
        $img->resize_gd( { width => 150 } );
        $img->save_jpeg('resized.jpg');
    
        # Easily resize artwork embedded within an audio file
        # You can use L<Audio::Scan> to obtain offset/length information
        my $img = Image::Scale->new( 'track.mp3', { offset => 2200, length => 34123 } );
        $img->resize_gd_fixed_point( { width => 75, height => 75, keep_aspect => 1 } );
        my $data = $img->as_png();

DESCRIPTION
    This module implements several resizing algorithms with a focus on low
    overhead, speed and minimal features. Algorithms available are:

      GD's copyResampled (floating-point)
      GD's copyResampled fixed-point (useful on embedded devices/NAS devices)
      GraphicsMagick's assortment of resize filters (floating-point)
      GraphicsMagick's Triangle filter in fixed-point

    Supported image formats include JPEG, GIF, PNG, and BMP for input, and
    JPEG and PNG for output.

    This module came about because we needed to improve the very slow
    performance of floating-point resizing algorithms on platforms without a
    floating-point unit, such as ARM devices like the SheevaPlug, and the
    Sparc-based ReadyNAS Duo. Previously it would take many seconds to
    resize using GD on the ReadyNAS but the conversion to fixed-point with a
    little assembly code brings this down to the range of well under 1
    second.

    Normal platforms will also see improvement, by removing all of the GD
    overhead this version of copyResampled is around 3 times faster while
    also using less memory.

    The fixed-point versions have an accuracy to around 4 decimal places so
    the quality of floating-point vs. fixed is essentially identical.

METHODS
  new( $PATH or \$DATA, [ \%OPTIONS ] )
    Initialize a new Image::Scale object from PATH, which may be any valid
    JPEG, GIF, PNG, or BMP file.

    Raw image data may also be passed in as a scalar reference. Using a file
    path is recommended when possible as this is more efficient and requires
    less memory.

    new() reads the image header, and will return undef if the header is
    invalid, so be sure to check for this.

    Optionally you can also pass in additional options in a hashref:

        offset
        length

    To access an image embedded within another file, such as an audio file,
    you can specify a byte offset and length.

  width()
    Returns the width of the original source image.

  height()
    Returns the height of the original source image.

  resize( \%OPTIONS )
    resize() uses the default resize algorithm, which is
    resize_gd_fixed_point. See below for details on the available options.

  resize_gd( \%OPTIONS )
  resize_gd_fixed_point( \%OPTIONS )
  resize_gm( \%OPTIONS )
  resize_gm_fixed_point( \%OPTIONS )
    The 4 resize methods available are:

        resize_gd - This is GD's copyResampled algorithm (floating-point)
        resize_gd_fixed_point - copyResampled (converted to fixed-point)
        resize_gm - GraphicsMagick, see below for filter options
        resize_gm_fixed_point - GraphicsMagick, only the Triangle filter is available in fixed-point mode

    Options are specified in a hashref:

        width
        height

    At least one of width or height are required. If only one is supplied
    the image will retain the original aspect ratio.

        filter

    For use with resize_gm() only. Choose from the following filters, sorted
    in order from least to most CPU time. This does not necessarily mean
    least to best quality, though! Be sure to do your own comparisons for
    quality.

        Point
        Box
        Triangle
        Hermite
        Hanning
        Hamming
        Blackman
        Gaussian
        Quadratic
        Cubic
        Catrom
        Mitchell
        Lanczos
        Bessel
        Sinc

    If no filter is specified the default is Lanczos if downsizing, and
    Mitchell for upsizing or if the image has an alpha channel.

        keep_aspect => 1

    Only useful when both width and height are specified. This option will
    keep the original aspect ratio of the source as well as center the image
    when resizing into a different aspect ratio. For best results, images
    altered in this way should be saved as PNG which will automatically add
    the necessary transparency around the image.

        bgcolor => 0xffffff

    When using keep_aspect, you can use bgcolor to define the background
    color of the padded portion of the image. Usually this should only be
    used if saving as JPEG because PNG will default to transparent. If this
    value is set and the image is saved as PNG, the PNG will not be
    transparent. The default bgcolor value is 0x000000 (black).

        ignore_exif => 1

    By default, if a JPEG image contains an EXIF tag with orientation info,
    the image will be rotated accordingly during resizing. To disable this
    feature, set ignore_exif to 1.

        memory_limit => $limit_in_bytes

    To avoid excess memory growth when resizing images that may be very
    large, you can specify this option. If the resize_*() method would
    result in a total memory allocation greater than $limit_in_bytes, the
    method will die. Be sure to wrap the resize call in an eval when using
    this option.

  save_jpeg( $PATH, [ $QUALITY ] )
    Saves the resized image as a JPEG to PATH. If a quality is not
    specified, the quality defaults to 90.

  as_jpeg( [ $QUALITY ] )
    Returns the resized JPEG image as scalar data. If a quality is not
    specified, the quality defaults to 90.

  save_png( $PATH )
    Saves the resized image as a PNG to PATH. Transparency is preserved when
    saving to PNG.

  as_png()
    Returns the resized PNG image as scalar data.

  jpeg_version()
  png_version()
  gif_version()
    Returns the version of the image library used. Returns undef if support
    for that image format was not built.

PERFORMANCE
    These numbers were gathered on my 2.4ghz MacBook Pro.

    JPEG image, 1425x1425 -> 200x200 (libjpeg v8 with scaling)

        GD copyResampled                        21.9/s
        resize_gm( { filter => 'Triangle' } )   65.7/s
        resize_gd_fixed_point                   67.9/s
        resize_gd                               69.4/s
        resize_gm_fixed_point                   74.5/s

    PNG image, 512x768 -> 200x133 (libpng 1.4.3)

        GD copyResampled                        14.7/s
        resize_gm( { filter => 'Triangle' } )   26.2/s
        resize_gm_fixed_point                   27.7/s
        resize_gd                               29.9/s
        resize_gd_fixed_point                   31.9/s

    Here are some numbers from a machine without floating-point support.
    (Marvell SheevaPlug 1.2ghz ARM9, JPEG 1425x1425 -> 200x200, libjpeg 6b
    with scaling)

        GD copyResampled                        1.08/s
        resize_gd                               2.16/s
        resize_gm( { filter => 'Triangle' } )   2.85/s
        resize_gd_fixed_point                   7.98/s
        resize_gm_fixed_point                   9.44/s

    And finally, from an even slower machine, the 240mhz Netgear ReadyNAS
    Duo which has extremely poor floating-point performance. (JPEG 1425x1425
    -> 200x200, libjpeg 6b with scaling)

        resize_gd                               0.029/s (34.5 s/iter)
        resize_gm( { filter => 'Triangle' } )   0.033/s (30.4 s/iter)
        resize_gd_fixed_point                   1.92/s  (0.522 s/iter)
        resize_gm_fixed_point                   2.07/s  (0.483 s/iter) (63x faster than floating-point!)

SEE ALSO
    TODO

AUTHOR
    Andy Grundman, <andy@hybridized.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2010 Andy Grundman

    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation; either version 2 of the License, or (at your
    option) any later version.



Changes
Revision history for Image:Scale

0.14    2017-11-27
        - Trying to resize certain kinds of corrupt JPEGs from an in-memory variable could get
          stuck in an infinite loop.

0.13    2016-11-24
        - save_jpeg() and save_png() will now properly stringify when passed something such as a
          Path::Tiny object.
        - Added back support for giflib 4.1.x.

0.12    2016-05-08
        - Skip JPEG binary comparison tests by default, as they are broken with libjpeg-turbo.
        - Correctly find libjpeg-turbo on Linux systems where some headers are installed
          in /usr/include/x86_64-linux-gnu
        - Removed hints/darwin.pl, don't try to be smarter than Perl about how to build things.

0.11    2015-10-07
        - Added $img->resized_height() and $img->resized_width() accessors that
          return the post-resize dimensions of the image. This can be useful if you
          only specified one dimension and need to know the other dimension that
          was calculated by the module. (RT #80269)
        - Changed ARM fixed_mul assembly code to not round the result, in order to
          match the output from x86 and C versions and pass the tests. (RT #93417)
        - Fixed GIF compilation when using giflib >= 5.0. (RT #106749)
        - Compilation fixes for newer OSX versions.
        - Point at new GitHub repo.

0.10    2015-08-23
        - GIF support now requires giflib >= 4.2.0, due to API changes.
          libungif is no longer supported. (RT #101679)
        - Fixed call to png_memcpy no longer working (RT #89927)

0.09    (unreleased)
        - Bug 17573, fixed several places where we could divide by 0.
        - Fixed JPEG version detection for libjpeg-turbo.
        - Fixed GIF library detection when using only --with-gif-includes switch.

0.08    2011-08-24
        - Forgot to add some new test files to the manifest.
        - Updated the Makefile and Win32 instructions to use libjpeg-turbo-1.1.1.

0.07    2011-08-24
        - Protect against resizing to a 0-width or 0-height image.
        - Use libungif if giflib is not found. (Max Soulé)
        - Fix build issue when building a non-x86 version using libjpeg-turbo's
          header file.
        - Hide orientation warning when EXIF orientation value is 0 (invalid but
          common for some reason).

0.06    2010-09-20 18:40:00
        - JPEG: Don't abort on fatal read errors if we have already processed
          some lines from the image. An image with blank space is better than
          no image at all.

0.05    2010-09-20 14:30:00
        - Faster resize when image doesn't have an alpha channel.
        - Faster JPEG compression when using libjpeg-turbo.

0.04    2010-09-17 15:40:00
        - More fixes for CPAN testers.

0.03    2010-09-16 18:00:00
        - Fixed compilation on Windows.

0.02    2010-09-16 17:00:00
        - Fixed crash when reading JPEG images with large Exif tags.
        - Improved fixed-point performance for x86.
        - Fixes for various test failures.

0.01    2010-09-08 23:40:00
        - Initial release.


Makefile.PL
use strict;

use 5.008;

use Config;
use ExtUtils::MakeMaker qw(WriteMakefile);
use File::Spec::Functions;
use Getopt::Long;

my (@INC, @LIBPATH, @LIBS);
my ($JPEG, $PNG, $GIF, $MYEXTLIB);
my ($JPEG_VERSION, $PNG_VERSION, $GIF_VERSION);

my ($help, $jpeg_inc, $jpeg_lib, $png_inc, $png_lib, $gif_inc, $gif_lib);
my ($jpeg_static, $png_static, $gif_static);

my $DEFINES = '-O2';
$DEFINES .= ' -Wall' unless $^O =~ /sun|solaris/i;
$DEFINES .= ' -Wno-unused-value -Wno-format-security -Winline -ffast-math' unless $^O =~ /Win32|sun|solaris/i;

if ( $^O !~ /Win32/i ) {
    # Check for Sparc ReadyNAS to enable ASM code
    if ( -e '/etc/raidiator_version' ) {
        if ( $Config{archname} eq 'sparc-linux' ) {
            $DEFINES .= ' -DPADRE';
        }
    }

    push @LIBPATH, '-L/usr/local/lib';
}

my $result = GetOptions(
    "help|h"               => \$help,
    "with-jpeg-includes=s" => \$jpeg_inc,
    "with-jpeg-libs=s"     => \$jpeg_lib,
    "with-jpeg-static"     => \$jpeg_static,
    "with-png-includes=s"  => \$png_inc,
    "with-png-libs=s"      => \$png_lib,
    "with-png-static"      => \$png_static,
    "with-gif-includes=s"  => \$gif_inc,
    "with-gif-libs=s"      => \$gif_lib,
    "with-gif-static"      => \$gif_static,
);

if ($help || !$result) {
    print STDERR <<END;
Usage: perl Makefile.PL [options]

Configure Image::Scale module.

Options:
    --with-jpeg-includes    Path to directory containing jpeglib.h
    --with-jpeg-libs        Path to directory containing libjpeg
    --with-jpeg-static      Statically link libjpeg
    --with-png-includes     Path to directory containing png.h
    --with-png-libs         Path to directory containing libpng
    --with-png-static       Statically link libpng
    --with-gif-includes     Path to directory containing gif_lib.h
    --with-gif-libs         Path to directory containing libgif
    --with-gif-static       Statically link libgif

END

    exit 0;
}

my @check = qw(/usr/include /usr/local/include /opt/local/include);

# Check in environment-supplied locations before standard ones
unshift @check, split /:/, $ENV{INCLUDE};

if ( $^O =~ /Win32/i ) {
    push @check, '../libjpeg-turbo-1.1.1';
    push @check, '../lpng143';
    push @check, '../giflib-4.1.6/lib';
}

# Look for libjpeg
for my $incdir ( $jpeg_inc, @check ) {
    if ( $incdir && -e catfile($incdir, 'jpeglib.h') ) {
        $JPEG = 1;
        $jpeg_inc = $incdir;

        unshift @INC, '-I' . $jpeg_inc;
        if ($jpeg_lib) {
            unshift @LIBPATH, '-L' . $jpeg_lib;
        }

        if ($jpeg_static) {
            if ($jpeg_lib) {
                $MYEXTLIB .= $jpeg_lib . '/' . 'libjpeg$(LIB_EXT) ';
            }
            else {
                $MYEXTLIB .= $jpeg_inc . '/../lib/libjpeg$(LIB_EXT) ';
            }
        }
        else {
            push @LIBS, '-ljpeg';
        }

        if ( $^O =~ /Win32/i ) {
            $JPEG_VERSION = _re_find( catfile($incdir, 'build', 'jconfig.h'), qr/JPEG_LIB_VERSION\s+(\d+)/ ) || 'unknown';
        }
        else {
            $JPEG_VERSION = _re_find( catfile($incdir, 'jpeglib.h'), qr/JPEG_LIB_VERSION\s+(\d+)/ ) || 'unknown';
            if ( $JPEG_VERSION eq 'unknown' ) {
                # Also check jconfig.h as it contains the version when using libjpeg-turbo
                if ( -d catdir($incdir, 'x86_64-linux-gnu') ) { # jconfig.h header is sometimes installed in a subdir
                    $JPEG_VERSION = _re_find( catfile($incdir, 'x86_64-linux-gnu', 'jconfig.h'), qr/JPEG_LIB_VERSION\s+(\d+)/ ) || 'unknown';
                }
                else {
                    $JPEG_VERSION = _re_find( catfile($incdir, 'jconfig.h'), qr/JPEG_LIB_VERSION\s+(\d+)/ ) || 'unknown';
                }
            }
        }

        $DEFINES .= " -DHAVE_JPEG -DJPEG_VERSION=$JPEG_VERSION";

        last;
    }
}

# Look for libpng
for my $incdir ( $png_inc, @check ) {
    if ( $incdir && -e catfile($incdir, 'png.h') ) {
        $PNG = 1;
        $png_inc = $incdir;

        unshift @INC, '-I' . $png_inc;
        if ($png_lib) {
            unshift @LIBPATH, '-L' . $png_lib;
        }

        if ($png_static) {
            if ($png_lib) {
                $MYEXTLIB .= $png_lib . '/' . 'libpng$(LIB_EXT) ';
            }
            else {
                $MYEXTLIB .= $png_inc . '/../lib/libpng$(LIB_EXT) ';
            }

            push @LIBS, '-lz';
        }
        else {
            push @LIBS, '-lpng -lz';
        }

        $PNG_VERSION = _re_find( catfile($incdir, 'png.h'), qr/PNG_LIBPNG_VER_STRING\s+"([\d\.]+)"/ ) || 'unknown';

        $DEFINES .= " -DHAVE_PNG -DPNG_VERSION=$PNG_VERSION";

        last;
    }
}

# Look for giflib
for my $incdir ( $gif_inc, @check ) {
    if ( $incdir && -e catfile($incdir, 'gif_lib.h') ) {
        $gif_inc = $incdir;

        unshift @INC, '-I' . $gif_inc;
        if ($gif_lib) {
            unshift @LIBPATH, '-L' . $gif_lib;
        }

        if ($gif_static) {
            if ($gif_lib) {
                $MYEXTLIB .= $gif_lib . '/' . "libgif\$(LIB_EXT) ";
            }
            else {
                $MYEXTLIB .= $gif_inc . "/../lib/libgif\$(LIB_EXT) ";
            }
        }
        else {
            push @LIBS, "-lgif";
        }

        my $GIF_41    = _re_find( catfile($incdir, 'gif_lib.h'), qr/GIF_LIB_VERSION " Version (4\.1), "/ );
        my $GIF_MAJOR = _re_find( catfile($incdir, 'gif_lib.h'), qr/GIFLIB_MAJOR (\d+)/ );
        my $GIF_MINOR = _re_find( catfile($incdir, 'gif_lib.h'), qr/GIFLIB_MINOR (\d+)/ );

        if ( $GIF_MAJOR ) {
            $GIF_VERSION = "${GIF_MAJOR}.${GIF_MINOR}";
            if ( $GIF_VERSION >= 4.2 ) {
                $GIF = 1;
                $DEFINES .= " -DHAVE_GIF -DGIF_VERSION=$GIF_VERSION";

                # giflib >= 5.1 changes DGifClose()
                if ( $GIF_VERSION >= 5.1 ) {
                    $DEFINES .= " -DGIFLIB_API_51";
                }

                # giflib >= 5.0 change DGifOpen(), GifErrorString(), GifAddExtensionBlock() and struct SavedImage
                if ( $GIF_VERSION >= 5.0 ) {
                    $DEFINES .= " -DGIFLIB_API_50";
                }
            }
        }
        elsif ( $GIF_41 ) {
            $GIF = 1;
            $GIF_VERSION = '4.1';
            $DEFINES .= ' -DHAVE_GIF -DGIF_VERSION=4.1 -DGIFLIB_API_41';
        }

        last;
    }
}

# Require at least one of jpeg or png
if ( !$JPEG && !$PNG ) {
    my $err = "Image::Scale requires at least one of libjpeg or libpng, please install one or both and try again.\n";

    if ( $^O =~ /Win32/i ) {
        $err .= "\nPlease read the README.win32 file for build instructions on Windows.\n";
    }

    warn $err;

    exit 0;
}

unshift @INC, '-I. -I.. -Isrc -Iinclude';

print "\n";
print "Image::Scale will be built with:\n\n";
print "  JPEG support:     " . ( $JPEG ? "yes ($jpeg_inc, libjpeg version $JPEG_VERSION)" : "no, install libjpeg (v8 recommended)" ) . "\n";
print "  PNG support:      " . ( $PNG ? "yes ($png_inc, libpng version $PNG_VERSION)" : "no, install libpng (v1.4.x recommended)" ) . "\n";
print "  GIF support:      " . ( $GIF ? "yes ($gif_inc, giflib version $GIF_VERSION)" : "no, install giflib v4.1 or higher" ) . "\n";
print "\n";

if ( $^O =~ /Win32/i ) {
    *MY::postamble = sub {};

    if ($JPEG) {
        push @INC, '-I../libjpeg-turbo-1.1.1/build'; # for jconfig.h
        $MYEXTLIB .= '../libjpeg-turbo-1.1.1/build/jpeg-static.lib ';
    }

    if ($PNG) {
        push @INC, '-I../zlib';
        $MYEXTLIB .= '../lpng143/libpng.lib ../zlib/zlib.lib ';
    }

    if ($GIF) {
        $MYEXTLIB .= '../giflib-4.1.6/windows/Release/giflib.lib ';
    }
}

my $inc_files = join(' ', glob 'include/*.h');
my $src_files = join(' ', glob 'src/*.c');

WriteMakefile(
    NAME              => 'Image::Scale',
    VERSION_FROM      => 'lib/Image/Scale.pm',
    PREREQ_PM         => { 'Test::NoWarnings' => 0 },
    ABSTRACT_FROM     => 'lib/Image/Scale.pm',
    AUTHOR            => 'Andy Grundman <andy@hybridized.org>',
    INC               => join(' ', @INC),
    LIBS              => [ join(' ', @LIBPATH, @LIBS) ],
    DEFINE            => $DEFINES,
    MYEXTLIB          => $MYEXTLIB,
    depend            => { 'Scale.c' => "$inc_files $src_files" },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/andygrundman/Image-Scale.git',
                web  => 'https://github.com/andygrundman/Image-Scale',
            },
        },
    },
    MIN_PERL_VERSION => 5.008,
    LICENSE => 'gpl_2',
);

sub _re_find {
    my ( $file, $re ) = @_;

    my $result;

    open my $fh, '<', $file or die "Unable to read $file\n";
    while ( my $line = <$fh> ) {
        if ( $line =~ $re ) {
            $result = $1;
            last;
        }
    }
    close $fh;

    return $result;
}
